Move Artifacts and BinaryPlatforms into Base

staticfloat opened the pull request:
This is the fundamental infrastructure needed for the 1.6 `Artifacts` rework.

* Imports `Pkg.BinaryPlatforms` to `Base.BinaryPlatforms`.  I didn't think it was useful enough to be exported from `Base`, so it's currently usable through `using Base.BinaryPlatforms`.  It has been more or less completely rewritten to be more flexible, more compiler-friendly and to power us into the age of https://github.com/JuliaLang/Pkg.jl/issues/1780

* Imports `Pkg.Artifacts` to its own stdlib, `Artifacts`.  Includes locating pieces of `Pkg.Artifacts`, not the installation pieces, which are still within `Pkg`.  Scripts that want to generate artifacts should still `using Pkg.Artifacts`, but scripts that just want to use artifacts that have already been generated are free to simply `using Artifacts`, which should cut down on time to first load by a nice chunk.  If a package attempts to `@artifact"foo"` on an artifact that does not exist, the `Artifacts` module will dynamically load `Pkg` and invoke the downloading routines through an inference barrier.

vtjnash commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+module BinaryPlatforms
+
+export AbstractPlatform, Platform, HostPlatform, platform_dlext, tags, arch, os,
```
```suggestion
export Platform, HostPlatform, platform_dlext, tags, arch, os,
```

Doesn't make sense to export a deprecated type. I'm also not sure other small names like `os` make sense to be exported either, or helper functions with internal-sounding names like `parse_dl_name_version`.

staticfloat replied:
While I agree with you, I think we need to keep backward-compatibility.  Should I instead only export these from `Pkg.BinaryPlatforms`?

KristofferC replied:
`BinaryPlatforms` was never an exposed user interface was it?

staticfloat replied:
It is as exposed now as it was in `Pkg`.  Packages can `using Pkg.BinaryPlatforms` and they get a bunch of stuff, just like they can `using Base.BinaryPlatforms`.  But neither are re-exported from `Base` or from `Pkg`.

KristofferC replied:
Ok, but it wasn't documented / exposed from Pkg either, right?

KristofferC replied:
I guess the jll files gotta use it though? Is that what we want to be backwards compatible with?

staticfloat replied:
Right; older JLLs, BB and BP are the main users of it I think.

KristofferC replied:
In my opinion, it would be good to know, what are the exact things that need to be exported to make old JLLs work and what are the exact features that the new JLLs need? These are the things we should not break in the future Julias. With the exception of that, I think almost everything else should be considered internal.

KristofferC replied:
> so only users that run using Base.BinaryPlatforms are going to see any of these names.

Ok, and `Base` doesn't export `BinaryPlatforms` so everything is there is subject to change anyway. Then you can indeed export whatever.

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+            "os" => os,
+        )
+        for (tag, value) in kwargs
+            tag = lowercase(string(tag))
```
```suggestion
            tag = lowercase(string(tag)::String)
```

am I getting this right @timholy?

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+            # set to `true`, it will cause an error later on.
+            if tag ∈ ("libgfortran_version", "libstdcxx_version", "os_version")
+                normver(x::VersionNumber) = string(x)
+                function normver(str::AbstractString)
```
Perhaps make this

```suggestion
                function normver(str::String)
```

and then call it with `normver(string(value)::String)` below?

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+            end
+
+            # Normalize and reject nonos
+            value = lowercase(string(value))
```
```suggestion
            value = lowercase(string(value)::String)
```

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+            tags["call_abi"] = "eabihf"
+        end
+
+        # If the user is asking for strict validation, do so.
```
I mean, that's what the code just below it says 🤷

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+struct Platform <: AbstractPlatform
+    tags::Dict{String,String}
+    # The "compare strategy" allows selective overriding on how a tag is compared
+    compare_strategies::Dict{String,Any}
```
```suggestion
    compare_strategies::Dict{String,Function}
```

or?

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+
+function validate_tags(tags::Dict)
+    throw_invalid_key(k) = throw(ArgumentError("Key \"$(k)\" cannot have value \"$(tags[k])\""))
+    ARCHITECTURE_FLAGS = Dict(
```
make this a const and move it outside?

staticfloat replied:
Is that preferable?  I thought it was better to have it embedded so that it doesn't show up in as many scopes.

KristofferC replied:
Otherwise, it will be created at every call, no? But maybe not performance sensitive so it doesn't matter? If you think about scope I think you would usually do it as


```
let
const FOO = ...
global function f(...)
    # use FOO
end
end
```

staticfloat replied:
Yeah, I kind of thought the compiler would notice it was constant at compile-time and would store the completed object.  It looks like it doesn't do that though.

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+# Simple equality definition; for compatibility testing, use `platforms_match()`
+Base.:(==)(a::AbstractPlatform, b::AbstractPlatform) = tags(a) == tags(b)
+
+function validate_tags(tags::Dict)
```
```suggestion
function validate_tags(tags::Dict{String, String})
```

I think?

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+    HostPlatform(parse(Platform, Base.BinaryPlatforms.host_triplet()))
+"""
+function Platform()
+    global default_host_platform
```
```suggestion
```

I think?

It seems a bit odd for me that this is called `Platform` but it isn't a constructor.

staticfloat replied:
Yes, I struggled with this.  I think we should perhaps instead call this `HostPlatform()`.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+"""
+    parse_toml(path::String)
+
+Uses Base.TOML to parse a TOML file
```
Feels like this doesn't really need a docstring? The reason to not add a dependency to the TOML stdlib is to keep this module as lean as possible? Could be added as a comment.

staticfloat replied:
Why would I use a comment instead of a docstring?  That seems like a strictly worse developer experience.  I'd much rather be able to look up this docstring in editors and the REPL.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+function with_artifacts_directory(f::Function, artifacts_dir::String)
+    try
+        ARTIFACTS_DIR_OVERRIDE[] = artifacts_dir
+        f()
```
```suggestion
        string(f())::String
```

maybe

staticfloat replied:
Why is `f()` returning a `String`?  I don't think that's true.

KristofferC replied:
I think I mixed this up with some other function...

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+current set of depot paths and the current artifact directory override via the method
+`with_artifacts_dir()`.
+"""
+function artifacts_dirs(args...)
```
Looks like this is only called with a `String` so far?

staticfloat replied:
Within this file, yes.  I could probably annotate `args` as `::AbstractString...`.

KristofferC replied:
So `args...` is really a path here? I don't get why it is accepting a vararg though, is it to not have the caller have to do a `joinpath`? Seems easier to read if it was just one argument and also likely better from the compiler point of view that you don't compile different versions for different number of arguments.

staticfloat replied:
Yeah, it was meant to mimic `abspath()` in that regard.  It does appear that I never ended up using that functionality, so I'll just restrict it to a single string.

staticfloat replied:
Aha, I figured out why I had this; it was actually as a convenient definition for `artifacts_dirs()`, which I use in `Pkg`.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+"""
+function artifacts_dirs(args...)
+    if ARTIFACTS_DIR_OVERRIDE[] === nothing
+        return [abspath(depot, "artifacts", args...) for depot in Base.DEPOT_PATH]
```
```suggestion
        return String[abspath(depot, "artifacts", args...) for depot in Base.DEPOT_PATH]
```

staticfloat replied:
```suggestion
        return String[abspath(depot, "artifacts", args...) for depot in Base.DEPOT_PATH]
```

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+        return [abspath(depot, "artifacts", args...) for depot in Base.DEPOT_PATH]
+    else
+        # If we've been given an override, use _only_ that directory.
+        return [abspath(ARTIFACTS_DIR_OVERRIDE[], args...)]
```
```suggestion
        return String[abspath(ARTIFACTS_DIR_OVERRIDE[], args...)]
```

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+        function parse_mapping(mapping::String, name::String)
+            if !isabspath(mapping) && !isempty(mapping)
+                try
+                    mapping = Base.SHA1(mapping)
```
there's no `tryparse` for `Base.SHA1`? Perhaps should be added?

staticfloat replied:
Agreed.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+        end
+
+        for (k, mapping) in depot_override_dict
+            # First, parse the mapping. Is it an absolute path, a valid SHA1-hash, or neither?
```
```suggestion
            # First, parse the mapping. Is it an absolute path, a valid SHA1-hash, or neither?
            mapping::Union{Dict{String, Any}, String}
```

I think?

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+                else
+                    overrides[:hash][hash] = mapping
+                end
+            elseif isa(mapping, Dict)
```
```suggestion
            elseif isa(mapping, Dict{String, Any})
```

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+                # For each name in the mapping, update appropriately
+                for name in keys(mapping)
+                    # If the mapping for this name is the empty string, un-override it
+                    if mapping[name] == ""
```
Why loop over `keys(mapping)` and then immediately lookup `mapping[name]`? Maybe just loop directly over the dict?

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+            elseif isa(mapping, Dict)
+                # Convert `k` into a uuid
+                uuid = try
+                    Base.UUID(k)
```
We need a `tryparse(UUID, k)`..

KristofferC commented on NEWS.md:
```diff
@@ -84,6 +84,12 @@ Standard library changes
 * `push!(c::Channel, v)` now returns channel `c`. Previously, it returned the pushed value `v` ([#34202]).
 * `RegexMatch` objects can now be probed for whether a named capture group exists within it through `haskey()` ([#36717]).
 * A new standard library `TOML` has been added for parsing and printing [TOML files](https://toml.io) ([#37034]).
+* The `Pkg.BinaryPlatforms` module has been moved into `Base` as `Base.BinaryPlatforms` and heavily reworked.
```
Was this ever documented / exposed to users? There is nothing in the Pkg docs about it I think.

staticfloat replied:
There are a few other references to it, in `Tectonic`, `BugReporting`, `PlotlyBase`, `NNlib`, and `MethodAnalysis`, but I agree it's not too important.  I mostly just wanted to have something to show for weeks of effort.  :P

KristofferC replied:
Show them `@time import Plots` instead or `@time import Gtk`?  :P Will likely impress move.

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+end
+
+# Other `Platform` types can override this (I'm looking at you, `AnyPlatform`)
+tags(p::Platform) = p.tags
```
Might want to use @timholy s "interface" trick like in 

https://github.com/JuliaLang/julia/commit/48fc8b4e5312eb9d815c7bfe665c55625a225654

here. Basically

```jl
function Base.getproperty(platform::AbstractPlatform, fieldname::Symbol)
    if fieldname === :tags
        return getfield(platform, :state)::Dict{String, String}
    else
        getfield(platform, fieldname)
    end
end
```

and then use `.tags` everywhere. That ensures inference know that `.tags` has to be a `Dict{String, String}`.

staticfloat replied:
I thought this kind of overriding of `getproperty()` was frowned upon?  Is that an outdated Julia style?

timholy replied:
This isn't about adding OOP-like behavior; it's really about fixing inference problems you have abstract types but can't infer the specific concrete subtype. (E.g., if you extract `dict[key]` when `dict::Dict{String,Platform}`, you know the result is a `Platform` but not the concrete subtype.) Inference will generally return `::Any` as the inferred type of field accesses to abstract types. Defining a `getproperty` lets you add the type-asserts to fix it. See https://timholy.github.io/SnoopCompile.jl/stable/snoopr/#Inferrable-field-access-for-abstract-types-1 for more explanation.

staticfloat replied:
I'm a little confused then because there are no abstract types here.  `Platform` is concrete, and `tags` is also concrete, it's a `Dict{String,String}`.  I don't want to define a `tags()` for other `AbstractPlatform` types because they will likely do something completely different (e.g. `AnyPlatform` won't have _any_ tags, it will just return a constant empty set, and other objects like the compatibility `Linux` type will sub out to a `tags()` on a wrapped `Platform` object).

timholy replied:
Then it doesn't matter. I was going by the current implementation for which
```julia
julia> isabstracttype(Pkg.BinaryPlatforms.Platform)
true
```

KristofferC replied:
>  I don't want to define a tags() for other AbstractPlatform types because they will likely do something completely different 

I assumed you would since there is a bunch of code that calls `tags` on `AbstractPlatform` e.g.

```
 Base.haskey(p::AbstractPlatform, k::String) = haskey(tags(p), k)
```

and

```
# Simple equality definition; for compatibility testing, use `platforms_match()`
Base.:(==)(a::AbstractPlatform, b::AbstractPlatform) = tags(a) == tags(b)
```

etc. Should these be restricted to `Platform` then?

staticfloat replied:
> Then it doesn't matter. I was going by the current implementation for which

Gotcha!  Makes sense.

> I assumed you would since there is a bunch of code that calls `tags` on AbstractPlatform

So the semantics of what `tags` does is stable; `tags(p::AbstractPlatform)` can be assumed to return a `Dict{String,String}`, but how it generates that dict is completely up to the `AbstractPlatform` object.  For an `AnyPlatform` object it will likely return an empty Dict.  For a `Linux` object (a compatibility wrapper for older BinaryProvider releases and such) it will sub out to an internal `Platform` object, etc...  So I still want to call `tags()` on other `AbstractPlatforms`, but I'm leaving it up to those other `AbstractPlatform` types to define just what needs to happen inside of `tags()`.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+    c_simple = "/path/to/c_simple_dir"
+    libfoo = "fb886e813a4aed4147d5979fcdf27457d20aa35d""
+
+This file defines four overrides; two which override specific artifacts identified
```
Referring to "this file" in a docstring is a bit non-ideal maybe. The function could be moved and the docstring (as opposed to a comment) is mostly supposed to be read in some other medium like in the terminal or html rendering where the concept of the file is just an implementation detail.

staticfloat replied:
It seems very straightforward to me what file it's referring to: the ``~/.julia/artifacts/Override.toml` file it was talking about in the paragraph before.

KristofferC replied:
I understand now.

KristofferC commented:
One general comment is that it is quite hard for me to understand what is the public API and what is internal.

For example, there is the function `split_artifact_slash` which has full documentation with docstrings and everything but it seems to just be called internally in another function. Is that really something that is going to be a public API and subject to non-breakage etc?

staticfloat commented:
> One general comment is that it is quite hard for me to understand what is the public API and what is internals.
> 
> For example, there is the function split_artifact_slash which has full documentation with docstrings and everything but it seems to just be called internally in another function. Is that really something that is going to be a public API and subject to non-breakage etc?

In my opinion, anything that is not completely trivial should have docstrings to aid others (and, honestly, myself) when editing this file in the future.  The demarcator for public/private isn't how well documented it is, it's whether it's exported or not.  Since both of these modules had certain exports in `Pkg`, I've defaulted to mostly exporting the same things from them (So that e.g. users can switch from `Pkg.Artifacts` to `Artifacts` with as little disruption as possible) but especially in the case of `BinaryPlatforms` there's not much need for that.

KristofferC commented:
>  The demarcator for public/private isn't how well documented it is, it's whether it's exported or not.

This is not really true. There are many things that are public that are not exported (and cannot be exported due to name clashes). For example `Profile.print`, `TOML.parse` etc. With few exceptions, only public-facing API in the Julia repo has docstrings.

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+            "os" => os,
+        )
+        for (tag, value) in kwargs
+            tag = lowercase(string(tag))
```
```suggestion
            tag = lowercase(String(tag)::String)
```

would make the compiler infer that `tag` is a `String` even though we don't know what `tag` is.

staticfloat replied:
Huh, I didn't realize that that was necessary.  Are there `String()` methods that don't return `String`?

timholy replied:
```julia
julia> Base.return_types(String, (Any,))
9-element Vector{Any}:
 String
 String
 String
 String
 Any
 String
 String
 String
 String

julia> f(c) = String(c[1])
f (generic function with 1 method)

julia> @code_typed f(Any[1])
CodeInfo(
1 ─ %1 = Base.arrayref(true, c, 1)::Any
│   %2 = Main.String(%1)::Any
└──      return %2
) => Any
```

timholy replied:
Though here I wonder if the compiler knows `tag` is a `Symbol` since it's a kwarg name.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+            # First, parse the mapping. Is it an absolute path, a valid SHA1-hash, or neither?
+            try
+                mapping = parse_mapping(mapping, k)
+            catch
```
What are the exceptions that can be thrown here? If it is only about the `Base.SHA1` call then I think it makes sense to return `nothing` from `parse_mapping` and switch on that instead of a catch all `catch` that would just give you a log message even if someone later made a typo inside `parse_mapping`. This exact thing happened at https://github.com/JuliaLang/Pkg.jl/pull/1993.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,634 @@
+                else
+                    overrides[:hash][hash] = mapping
+                end
+            elseif isa(mapping, Dict{String, Any})
```
maybe make this into an `else` and type assert or an `else` and error. It just feels a little bit scary that we maybe missed that `mapping` is some other type and that it would then just go through and do nothing?

staticfloat replied:
Considering that `mapping` is coming from `parse_mapping`, I think the types are very tightly controlled.  But I believe in myself less than the compiler, so I still think it's a good idea.  :)

staticfloat commented:
> This is not really true. There are many things that are public that are not exported (and cannot be exported due to e.g. name clashes). For example Profile.print, TOML.parse etc.

These are exceptions due to name clashes, it's not good practice to have public APIs that are unexported.  In fact, it is then even more important to have documentation stating that `TOML.parse` is a public API.

> Also, in packages, you might now want to export something that has a very usual name that is likely to be too generic to be understandable on its own.

I'm not sure what you're saying here; but naming of documented, non-exported symbols can be changed before exporting.

> What matters for public/private is documentation. With few exceptions, only public-facing API in the Julia repo has docstrings.

I strongly disagree.  While it may be the case that developers document public APIs more than internal ones, I argue that's a bad situation we accept because of limited developer time, not an ideal to be strived for.  We should never look down on more comments and documentation to help understand code.  Documentation is a tool to help people understand code, it is never desirable to make code less easy to understand.

timholy commented:
FWIW I like docstrings on internal methods too.

giordano commented:
I heard a few times the claim that public methods are those that have the docstring.  Don't we really have a better way to identify public and internal methods than having or not a docstring?

timholy commented:
Yeah, it seems silly to distinguish them that way. When I write Documenter docs, I only "publish" the ones I expect people to call, and I ignore the warnings about those internal functions that I've written a docstring for but which are really the internal interface.

KristofferC commented:
> These are exceptions due to name clashes, it's not good practice to have public APIs that are unexported. 

Where does this notion come from? `export` is used to put something in the global namespace which is a shared resource and as you use more packages it will inherently start erroring. Many packages break as soon as we export something more from Base.
Also, for code readers, it makes it impossible to locally see where a symbol is coming from without help from an IDE which is awful when you are trying to get familiar with a new codebase. Anyway, this is a side discussion.


> I strongly disagree. While it may be the case that developers document public APIs more than internal ones, I argue that's a bad situation we accept because of limited developer time, not an ideal to be strived for. We should never look down on more comments and documentation to help understand code. Documentation is a tool to help people understand code, it is never desirable to make code less easy to understand.

When I said "documentation determines public/private", I should have been more specific and said, "public-facing documentation determines public/private". Adding a docstring to something indeed doesn't make it public, putting it into the manual does.

KristofferC commented:
I mean, the reason why I bring this up is that I see quite a few places which e.g. could have problems with invalidations (and one of the main reasons this code movement is being done is AFAIU due to compilation time). However, I don't want to nitpick every single place when it might not be a problem in practice. Therefore, if we have the freedom to change things, then it is probably better to merge, and to these fixes later.

But now e.g. `parse_dl_name_version(path::AbstractString, platform::AbstractPlatform)` is exported and documented so if we find that there are a bunch of invalidations from that and want to tightent it to `Union{String, SubString{String}}` are we then breaking the interface? It's exported so maybe? But the module itself is not so maybe not? But the jll packages will use `BinaryPlatforms` so I guess it is effectively public. But the jll packages might only call this with `String`. So is it ok to change? Etc.

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -0,0 +1,642 @@
+
+# Support `AbstractString`s, but avoid compilers needing to track backedges for callers
+# of these functions in case a user defines a new type that is `<: AbstractString`
+parse_toml(path::AbstractString) = parse_toml(string(path))
```
```suggestion
parse_toml(path::AbstractString) = parse_toml(string(path)::String)
```

are needed.

But perhaps it is enough to restrict to `Union{String, SubString{String}}`?

KristofferC commented on base/binaryplatforms.jl:
```diff
@@ -0,0 +1,999 @@
+"""
+call_abi(p::AbstractPlatform) = get(tags(p), "call_abi", nothing)
+
+const platform_namess = Dict(
```
Typo I think?

Assuming that there is no intention to dynamically add platforms, `platform_name` below could be written as

```
function platform_name(p::AbstractPlatform)
    n = os(p)
    n === nothing && return "Unknown"
    n == "linux"   ? "Linux"   :
    n == "macos"   ? "macOS"   :
    n == "windows" ? "Windows" :
    n == "freebsd" ? "FreeBSD" :
    error("unhandled os $(repr(n))")
end
```

which avoids having to have the dictionary at all, makes it clear that `platform_namess` is not something that is extended and showes that there is a possible unrecoverable error when something unexpected is returned from `os`. The real hardcore version would be to define with an `@enum PlatformType` the different platforms listed here and make `os` return a `Union{Nothing, PlatformType}`  thus making the error impossible by the type system.

staticfloat replied:
> Typo I think?

Yes, thank you!

> Assuming that there is no intention to dynamically add platforms

I think having this escape hatch open is prudent.  While I don't expect to add any more future platforms in the near future, who knows, perhaps we'll add OpenBSD support or something and then we'll need to modify this.  It would be a lot easier if our JLL packages can push something into this dictionary via a Pkg hook or something.  :)

KristofferC commented:
Anyway, I didn't mean for my attempt at a review to start this type of a "heated" side discussion about stuff that is not directly related to the code in the PR. I just commented on some of my thoughts during reading the code and it was probably mostly a reflection of me being unfamiliar with the code and how it is being used in combination with that it didn't really use the standard documentation style as the rest of Base / stdlibs. In hindsight, I should probably have taken a bit different strategy.

The best way forward is probably to try to get stuff merged without too much nitpicking, get stuff in place, regenerate some jlls that can be used to test things out on, and then do eventual adjustments based on that.

staticfloat committed:
Add `BinaryPlatforms` module in `Base`

Adds a new `Platform` object that represents a computing platform as a
collection of tags mapping names to values.  Typical tags are things
such as processor `arch`, operating system, which libc is being used on
the platform, etc...  However more specific tags such as what version of
libgfortran the current Julia is linked against or what version of CUDA
is supported by the host are also representable.

Certain tags may require custom comparisons (such as Julia version
number, which should match using only the major and minor version
numbers) and so custom comparators are supported.

staticfloat committed:
Add `Artifacts` as a standard library

Imports `Pkg.Artifacts` into its own stdlib, Artifacts.  Includes the loading
and locating pieces of `Pkg.Artifacts`, but not the pieces for downloading and
installing artifacts, which still reside within Pkg.  Scripts that want to
generate artifacts should still use `Pkg.Artifacts`, but scripts that just want
to use artifacts that have already been generated are free to simply
`using Artifacts`, which should cut down on time to first load by a nice chunk.
If a package attempts to `@artifact"foo"` on an artifact that does not exist
the `Artifacts` module will dynamically load `Pkg` and invoke the downloading
routines through an inference barrier.

Also adds two new (minor) features:

* Slash-indexing: by popular demand, supports indexing into files within
  artifacts in one convenient syntax: `@artifact"FFMPEG/bin/ffmpeg"`
  will automatically perform the `joinpath()` for you after looking up
  the `FFMPEG` artifact.

* Compile-time constancy; when `@artifact_str` is given a constant
  `String`, it optimizes the generated code by looking up the
  appropriate hash at compile-time and embedding just that hash into the
  resultant code, saving on both lookup time and objects that must be
  serialized to the eventual precompilation `.ji` file.

staticfloat commented:
Alright, I've dropped the commit that bumps `Pkg`; this is now ready to merge.  Once this is merged, I'll merge [the PR against `Pkg`](https://github.com/JuliaLang/Pkg.jl/pull/2001), then we can bump `Pkg` in a future PR which will switch Julia over to a `Pkg` version that drops its duplicate `Artifacts` and `BinaryPlatforms` code.

stevengj commented:
a `git bisect` seems to indicate that this PR is triggering JuliaLang/IJulia.jl#968 … any ideas?
